RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -358,38 +529,258 @@ void TrackExtenderWithMTDT<TrackCollection>::produce(edm::Event& ev, const edm::
+                                           const Propagator* thePropagator,
+                                           TrajectoryStateClosestToBeamLine& tscbl) {
+    // get the state closest to the beamline
+    TrajectoryStateOnSurface stateForProjectionToBeamLineOnSurface =
+        traj.closestMeasurement(GlobalPoint(bs.x0(), bs.y0(), bs.z0())).updatedState();
+
+    if
+      UNLIKELY(!stateForProjectionToBeamLineOnSurface.isValid()) {
+        edm::LogError("CannotPropagateToBeamLine")
+            << "the state on the closest measurement isnot valid. skipping track.";
+        return false;
+      }
+
+    const FreeTrajectoryState& stateForProjectionToBeamLine = *stateForProjectionToBeamLineOnSurface.freeState();
+
+    TSCBLBuilderWithPropagator tscblBuilder(*thePropagator);
+    tscbl = tscblBuilder(stateForProjectionToBeamLine, bs);
+
+    if
+      UNLIKELY(!tscbl.isValid()) { return false; }
+
+    return true;
+  }
+
+  bool trackPathLength(const Trajectory& traj,
+                       const reco::BeamSpot& bs,
+                       const Propagator* thePropagator,
+                       double& pathlength) {
+    pathlength = 0.;
+
+    TrajectoryStateClosestToBeamLine tscbl;
+    bool tscbl_status = getTrajectoryStateClosestToBeamLine(traj, bs, thePropagator, tscbl);
+
+    if (!tscbl_status)
+      return false;
+
+    bool validpropagation = true;
+    double pathlength1 = 0.;
+    double pathlength2 = 0.;
+
+    //add pathlength layer by layer
+    for (auto it = traj.measurements().begin(); it != traj.measurements().end() - 1; ++it) {
+      const auto& propresult = thePropagator->propagateWithPath(it->updatedState(), (it + 1)->updatedState().surface());
+      double layerpathlength = std::abs(propresult.second);
+      if (layerpathlength == 0.) {
+        validpropagation = false;
+      }
+      pathlength1 += layerpathlength;
+    }
+
+    //add distance from bs to first measurement
+    if (traj.direction() == alongMomentum) {
+      const auto& propresult2 =
+          thePropagator->propagateWithPath(tscbl.trackStateAtPCA(), traj.firstMeasurement().updatedState().surface());
+      pathlength2 = propresult2.second;
+      if (pathlength2 == 0.) {
+        validpropagation = false;
+      }
+      pathlength = pathlength1 + pathlength2;
+    } else {
+      const auto& propresult2 =
+          thePropagator->propagateWithPath(tscbl.trackStateAtPCA(), traj.lastMeasurement().updatedState().surface());
+      pathlength2 = propresult2.second;
+      if (pathlength2 == 0.) {
+        validpropagation = false;
+      }
+      pathlength = pathlength1 + pathlength2;
+    }
+
+    return validpropagation;
+  }
+
+  struct TrackTofPidInfo {
+    TrackTofPidInfo(const double& magp2,
+                    const double& length,
+                    const double& t_mtd,
+                    const double& t_mtderr,
+                    const double& t_vtx,
+                    const double& t_vtx_err,
```

Comment:
args can be all passed by value here.

Structs should normally be just a set of named data members without fancy algorithms for filling the values.
Consider moving this constructor out into a "free" function